### DISCUSSION TITLE
Color replacement (in HSV?)

### DISCUSSION THREAD

**Discussion by ferropasha** [ImageMagick/ImageMagick#6503] [Help]
### ImageMagick version

7.0.9-1

### Operating system, version and so on

Windows 10

### Description

**I need advice on how to achieve better results with color replacement.**
The problem I am trying to solve is this. I have a number of scanned images of old newspaper pages, which I need to process in opencv (basically find page borders).  See example (actual page content blurred out). As you can see pages were scanned on uneven whitish-grey background cardboard.
![original](https://github.com/ImageMagick/ImageMagick/assets/131873751/3cbcb9a9-d404-4ee8-8007-53893456fbec)
Before feeding them to opencv I want to perform a bit of preprocessing on the pages in Imagemagick, **first of all replace background color with black**, because border finding algorithm presumes light colored pages on dark background,   
My first attempt was to use simple command like  **-fuzz 10% -fill black -opaque white** See example below
![bad_result_im](https://github.com/ImageMagick/ImageMagick/assets/131873751/9c2b7348-1423-456b-9e61-6c58e1d889c6)
The problem is cardboard background was unevenly lit in the lower part of the image, so this command produces **large white flare-like areas** in places where background color was much closer to grey than white, which messes up opencv algorithm.
By chance I stumbled upon a color replacing plugin for paint.net, which does more or less what I want. Replaced background still contains a lot of white, but it is less concentrated and opencv can live with it. Apparently this **plugin works in HSV colorspace**. See example below. What I need help with is **if it's possible to achieve the same result with Imagemagick and how?**
![desired_result_paintnet](https://github.com/ImageMagick/ImageMagick/assets/131873751/15d25149-5f0b-488d-bc85-1762961da4d4)

**Comment by snibgo** [comment]
My advice: put any JPG scans in the bin. JPG is only 8 bits/channel, with lossy compression that introduces noise, which makes any processing difficult.

You scan nearly-white objects against a nearly-white background, and then want to separate the two colours. Duh. Using a coloured background (eg blue or green) would make the job much simpler.

The foreground is slightly reddish, and the background is slightly blueish, so this mostly separates them:
```
magick ferroInp.jpg -fx "r>b ? 1 : 0" x.png
```

**Reply by ferropasha** [reply]
Thank you for your time. Your suggestion works but for a large image it is very slow. To elaborate a bit on my predicament

1) Actual scans are in 300dpi tif. I just converted example to jpg so it would be easier to upload here. Real originals are something like 120 to 150 mb per page.
2) I realize it would be a lot simpler to scan on a different background, but sadly I am just a guy assigned to deal with the consequences. We already have several thousands of images like this. Also there was a semi-legitimate reason to use such a background. Newspapers are printed on rather thin paper, so if you scan on dark background the color of the output changes perceptibly. 
3) **The main problem with background is not that it's white, but very unevenly white.** Actual whitish areas recolor nicely in IM with fuzz of say 8 to 10% without encroaching too much on the page itself.  The real problem is areas of greyish-blue colors at the bottom (speaking frankly I don't even know if it's shadow resulting from uneven lighting on the scanner or actual smudges of dust and newspaper ink on the cardboard background) that mess up the whole thing, **To recolor them I have to set fuzz to such a level that it practically obliterates the page altogether. I wonder how is it possible that this paint.net HSV thing can more or less deal with it and IM can't?**. If it could be of any value this is a screenshot of settings for the plugin, that produce desired result 
![Screen](https://github.com/ImageMagick/ImageMagick/assets/131873751/16b71bd0-e27e-4a96-93cb-31eaa3d4479b)
3) Does it make any difference what colorspace I use while recoloring in IM? If I attempt to do this in HSV how should I go about it?  **"-colorspace HSV -fuzz 10% -fill black -opaque white"** doesn't seem to do anything

**Comment by snibgo** [comment]
What version of IM do you have? `-fx` performance increased greatly about 1.5 years ago.

Your screenshot shows you are not detecting "white", but RGB=239,240,241 (out of 255, I assume). This is reasonable. If I crop to an area to the right of the newspaper, I get:
```
magick c:\web\im\ferroInp.jpg -crop 550x4090+2250+162 -scale "1x1^!" txt:
# ImageMagick pixel enumeration: 1,1,0,255,srgb
0,0: (232,238,237)  #E8EEED  srgb(90.812%,93.4521%,92.9827%)
```
So this average is about the same, RGB=232,238,237.

I don't know what the paint.net "colour detection method H,S V" is. Perhaps it is documented somewhere. Your image shows it has found false foreground and false background.

You can convert the image, and the color you are searching for, and the replacement color, to any colorspace. I don't know if that would help. I doubt that any simple colour test would succeed.

Do you really want the edges of the scanned paper? Perhaps you really want the edges of the *printed matter*, plus a small margin. That would be easier to detect, as the printed matter (text and illustrations) would be much darker than the near-white paper or the near-white background. A simple threshold might isolate the printed matter, then make the rest white, perhaps removing the thin dark vertical lines on the left by morphology. Find the bounding box of the printed matter by `-trim` or `-format %@`, and back off the bounds by 20 pixels or whatever.

If you do need the edges of the scanned paper, you could start from the edges of the printed matter, then search outwards until you hit "white".

**Reply by ferropasha** [reply]
My IM version is 7.0.9-1. 
I understand that no algorithm can separate background and foreground  in entirely reliable way in my case. My original idea was to use the images with recolored background as derivatives just to acquire page ROI coordinates, which would be later used to work with unaltered originals. That's why I don't mind if I get some false results in the foreground and background unless there are so many of them as to make finding ROI impossible.
The code for paint.net plugin (c# as I understand, fairly well commented) can be found in the first post here https://forums.getpaint.net/topic/24584-recolour-choice/
Not that I pretend to understand much of it but I think in HSV mode it converts target and replacement colors to HSV space then performs some calculations with them that I don't really comprehend the purpose of, then converts back to RGB
These are the parts I think are most relevant 
**Variables to store values acquired from GUI controls**

> ColorBgra Amount2 = ColorBgra.FromBgr(120,124,172); // colour to detect
bool Amount3 = true; // [0,1] include transparency (alpha) in detection
int Amount4 = 255; // [0,255] detection  transparency
double Amount5 = 10; // [0,100] detection tolerance %
ColorBgra Amount6 = ColorBgra.FromBgr(255,0,0); // replacement colour
int Amount7 = 255; // [0,255] replacement transparency
double Amount8 = 0; // [0,100] blend edges
bool Amount9 = false; // [0,1] keep unaltered detected only

**Values calculated for every pixel**

> cp = src[x,y];// cp = current pixel = set to source pixel

		 B = cp.B;G = cp.G;R = cp.R;A = cp.A;// source pixel

		 B1 = (int)Amount2.B;G1 = (int)Amount2.G;R1 = (int)Amount2.R;// colour to detect
		 A1 = Amount4; // detection transparency from slider - not from primary

		 t1 = Amount5/100;	//detection tolerance
		 t1b = t1 * 255;// t1 byte, from 0 to 255 but a double for easier maths!
		 t1p = t1 * 100;// t1 percent, from 0 to 100
		 t1d = t1 * 360;// t1 degrees, from 0 to 360
		 t1e = t1 * 441.67;// Euclidean range
		 t1l = t1 * 765;// tone/ lum range (255 * 3)
		 t2 = Amount8/100;	//blend tolerance


		 B2 = (int)Amount6.B;G2 = (int)Amount6.G;R2 = (int)Amount6.R;// replacement colour from color wheel
		 A2 = (int)Amount7;// replacement transparency from slider


		 Color col = cp.ToColor();// from Boltbait's tutorial part4 - thanks BB!
		 HsvColor hsv = HsvColor.FromColor(col);// source HSV color
		 int H = hsv.Hue;// from 0 to 360 (Red,Orange,Yellow,Green,Turquoise,Blue,Magenta,Carmine)
		 int S = hsv.Saturation;// from 0(grey) to 100 (full colour)
		 int V = hsv.Value;// from 0 (black) to 100 (white)
		 Color col1 = Amount2.ToColor();
		 HsvColor hsv1 = HsvColor.FromColor(col1);// detection HSV color
		 int H1 = hsv1.Hue;
		 int S1 = hsv1.Saturation;
		 int V1 = hsv1.Value;
		 double blend1 = 1;// have to assign a value
		 double blend;
		 double iblend;
		 int Adiff = Math.Abs(A - A1);// declare here as used in many cases
		 double Hdiff = Math.Abs(H - H1);
		 double Sdiff = Math.Abs(S - S1);
		 double Vdiff = Math.Abs(V - V1);

**Calculations in HSV mode**

> if ((Hdiff < t1d)  &&(Sdiff < t1p)  &&(Vdiff < t1p))
				{ if(!Amount3 ||(Amount3 && Adiff < t1b))
				   {
					blend1 = t2 * ((Hdiff / t1d) + (Sdiff / t1p) + (Vdiff / t1p));
					}
				}

**Converting back to RGB**

>  if(blend1 > 1){blend1 = 1;}// to prevent colour reversal if more than 1
		 blend = 1 - blend1;		// as in cases 1 & 7
		 iblend = 1 - blend;
		 B = (int)((B2 * blend) + (B * iblend));
		 G = (int)((G2 * blend) + (G * iblend));
		 R = (int)((R2 * blend) + (R * iblend));
		 A = (int)((A2 * blend) + (A * iblend));

**Comment by snibgo** [comment]
> My IM version is 7.0.9-1.

That was before `-fx` performance improved. If you use `-fx`, I suggest you upgrade to the current version of IM.

The code in https://forums.getpaint.net/topic/24584-recolour-choice/ seems to regard a hue of 1 degree as very different to a hue of 359 degrees, when in fact they are very close.
